Kernel/Arch/Processor.h

```diff
@@ -8,18 +8,201 @@
 #pragma once
 
 #include <AK/Function.h>
+#include <Kernel/Arch/CPUID.h>
 #include <Kernel/Arch/DeferredCallEntry.h>
+#include <Kernel/Arch/DeferredCallPool.h>
+#include <Kernel/Arch/FPUState.h>
+#include <Kernel/Arch/ProcessorSpecificDataID.h>
+#include <Kernel/Memory/VirtualAddress.h>
+
+#if ARCH(X86_64)
+#    include <Kernel/Arch/x86_64/DescriptorTable.h>
+#endif
 
 namespace Kernel {
 
-// FIXME: Move the InterruptsState enum and related functions inside the Processor class.
 enum class InterruptsState {
     Enabled,
     Disabled
 };
 
-InterruptsState processor_interrupts_state();
-void restore_processor_interrupts_state(InterruptsState);
+namespace Memory {
+class PageDirectory;
+}
+
+struct TrapFrame;
+class Thread;
+
+class Processor;
+
+extern Atomic<u32> g_total_processors;
+
+extern FPUState s_clean_fpu_state;
+
+// context_first_init is an architecture-specific detail with various properties.
+// All variants eventually call into the common code here.
+void do_context_first_init(Thread* from_thread, Thread* to_thread);
+extern "C" void exit_kernel_thread(void);
+extern "C" void thread_context_first_enter(void);
+extern "C" void do_assume_context(Thread* thread, u32 flags);
+extern "C" FlatPtr do_init_context(Thread* thread, u32) __attribute__((used));
+
+template<typename ProcessorT>
+class ProcessorBase {
+    friend class Processor;
```

Comment:
we should not need to friend Processor here. Just static_cast<ProcessorT>(*this) or such.